sql: interval division with float rounds differently from Python & PostgreSQL

see https://github.com/cockroachdb/cockroach/issues/66118#issuecomment-856366411 for solution

----


**Describe the problem**

Interval division with a float rounds differently from Python & PostgreSQL.

**To Reproduce**

```sql
CREATE TABLE "test1" (
    "id" serial NOT NULL PRIMARY KEY,
    "duration" interval NOT NULL
);

INSERT INTO "test1" ("duration") VALUES ('0 days 0.253000 seconds'::interval);

SELECT ("duration" / 3.2) FROM "test1";
     ?column?
-------------------
  00:00:00.079063
```

**Expected behavior**

Expected result of SELECT query: `00:00:00.079062`.

**Environment:**
 - CockroachDB version: 21.1.1 (and older versions)

**Additional context**

A Django test fails:
```
======================================================================
FAIL: test_durationfield_multiply_divide (expressions.tests.FTimeDeltaTests) (expr=3.2)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/tim/code/django/tests/expressions/tests.py", line 1565, in test_durationfield_multiply_divide
    self.assertEqual(
AssertionError: datetime.timedelta(microseconds=79063) != datetime.timedelta(microseconds=79062)
```

--- a/pkg/sql/logictest/testdata/logic_test/datetime
+++ b/pkg/sql/logictest/testdata/logic_test/datetime
@@ -1522,8 +1522,8 @@ ORDER BY
 ----
 01:00:00               00:30:00          02:00:00                00:30:00          02:00:00                04:14:01.320914                 00:14:10.32
 1 day                  12:00:00          2 days                  12:00:00          2 days                  4 days 05:36:31.701948          05:40:07.68
-1 mon                  15 days           2 mons                  15 days           2 mons                  4 mons 7 days 00:15:51.058425   7 days 02:03:50.4
-1 mon 2 days 04:00:00  16 days 02:00:00  2 mons 4 days 08:00:00  16 days 02:00:00  2 mons 4 days 08:00:00  4 mons 15 days 28:24:59.745978  7 days 14:20:47.04
+1 mon                  15 days           2 mons                  15 days           2 mons                  4 mons 7 days 00:15:51.0912     7 days 02:03:50.4
+1 mon 2 days 04:00:00  16 days 02:00:00  2 mons 4 days 08:00:00  16 days 02:00:00  2 mons 4 days 08:00:00  4 mons 15 days 28:24:59.778753  7 days 14:20:47.04
 
 subtest tz_utc_normalization
 

--- a/pkg/sql/logictest/testdata/logic_test/interval
+++ b/pkg/sql/logictest/testdata/logic_test/interval
@@ -367,3 +367,28 @@ subtest regression_62369
 
 query error "10000000000000000000000000000000000": value out of range
 SELECT INTERVAL '10000000000000000000000000000000000 year'
+
+query T
+SELECT i / 2 FROM ( VALUES
+  ('0 days 0.253000 seconds'::interval),
+  (INTERVAL '0.000001'::interval),
+  (INTERVAL '0.000002'::interval),
+  (INTERVAL '0.000003'::interval),
+  (INTERVAL '0.000004'::interval),
+  (INTERVAL '0.000005'::interval),
+  (INTERVAL '0.000006'::interval),
+  (INTERVAL '0.000007'::interval),
+  (INTERVAL '0.000008'::interval),
+  (INTERVAL '0.000009'::interval)
+) regression_66118(i)
+----
+00:00:00.1265
+00:00:00
+00:00:00.000001
+00:00:00.000002
+00:00:00.000002
+00:00:00.000002
+00:00:00.000003
+00:00:00.000004
+00:00:00.000004
+00:00:00.000004

--- a/pkg/util/duration/duration.go
+++ b/pkg/util/duration/duration.go
@@ -617,14 +617,37 @@ func (d Duration) MulFloat(x float64) Duration {
 
 // DivFloat returns a Duration representing a time length of d/x.
 func (d Duration) DivFloat(x float64) Duration {
-	monthInt, monthFrac := math.Modf(float64(d.Months) / x)
-	dayInt, dayFrac := math.Modf((float64(d.Days) / x) + (monthFrac * DaysPerMonth))
+	// In order to keep it compatible with PostgreSQL, we use the same logic.
+	// Refer to https://github.com/postgres/postgres/blob/e56bce5d43789cce95d099554ae9593ada92b3b7/src/backend/utils/adt/timestamp.c#L3266-L3304.
+	month := int32(float64(d.Months) / x)
+	day := int32(float64(d.Days) / x)
+
+	remainderDays := (float64(d.Months)/x - float64(month)) * DaysPerMonth
+	remainderDays = secRoundToEven(remainderDays)
+	secRemainder := (float64(d.Days)/x - float64(day) +
+		remainderDays - float64(int64(remainderDays))) * SecsPerDay
+	secRemainder = secRoundToEven(secRemainder)
+	if math.Abs(secRemainder) >= SecsPerDay {
+		day += int32(secRemainder / SecsPerDay)
+		secRemainder -= float64(int32(secRemainder/SecsPerDay) * SecsPerDay)
+	}
+	day += int32(remainderDays)
+	microSecs := float64(time.Duration(d.nanos).Microseconds())/x + secRemainder*MicrosPerMilli*MillisPerSec
+	retNanos := time.Duration(int64(math.RoundToEven(microSecs))) * time.Microsecond
 
 	return MakeDuration(
-		int64((float64(d.nanos)/x)+(dayFrac*float64(nanosInDay))),
-		int64(dayInt),
-		int64(monthInt),
+		retNanos.Nanoseconds(),
+		int64(day),
+		int64(month),
 	)
+}
+
+// secRoundToEven rounds the given float to the nearest second,
+// assuming the input float is a microsecond representation of
+// time
+// This maps to the TSROUND macro in Postgres.
+func secRoundToEven(f float64) float64 {
+	return math.RoundToEven(f*MicrosPerMilli*MillisPerSec) / (MicrosPerMilli * MillisPerSec)
 }
 
 // normalized returns a new Duration transformed using the equivalence rules.

--- a/pkg/util/duration/duration_test.go
+++ b/pkg/util/duration/duration_test.go
@@ -498,6 +498,18 @@ func TestFloatMath(t *testing.T) {
 			Duration{Months: 2, Days: 34, nanos: nanosInHour * 4},
 			Duration{Days: 23, nanos: nanosInHour * 13},
 		},
+		{
+			Duration{Months: 0, Days: 0, nanos: nanosInSecond * 0.253000},
+			3.2,
+			Duration{Months: 0, Days: 0, nanos: nanosInSecond * 0.8096},
+			Duration{Months: 0, Days: 0, nanos: nanosInSecond * 0.079062},
+		},
+		{
+			Duration{Months: 0, Days: 0, nanos: nanosInSecond * 0.000001},
+			2.0,
+			Duration{Months: 0, Days: 0, nanos: nanosInSecond * 0.000002},
+			Duration{Months: 0, Days: 0, nanos: nanosInSecond * 0},
+		},
 	}
 
 	for i, test := range tests {

